src/skmultiflow/anomaly_detection/half_space_trees.py

```diff
@@ -0,0 +1,548 @@
+                    anomaly_score = self._anomaly_score(node.left, X)
+
+        return anomaly_score
+
+    def build_tree(self, min, max, current_depth=0):
+        """
+        Builds a single half space tree.
+
+        Parameters
+        ----------
+        min: Array of floats
+            Minimum work range for every dimension.
+        max: Array of floats
+            Maximum work range for every dimension.
+        current_depth:
+            Indicates the current depth of the tree.
+        Returns
+        -------
+        HalfSpaceTreeNode:
+            root of the tree.
+        """
+        if self.max_depth == current_depth:
+            return HalfSpaceTreeNode(depth=current_depth, internal_node=False)
+        else:
+            random_dimension = random.randint(0, self.dimensions - 1)
+            p = (min[random_dimension] + max[random_dimension]) / 2.0
+            temp = max[random_dimension]
+            max[random_dimension] = p
+            left = self.build_tree(min, max, current_depth + 1)
+            max[random_dimension] = temp
+            min[random_dimension] = p
+            right = self.build_tree(min, max, current_depth + 1)
+            return HalfSpaceTreeNode(left=left, right=right, split_attribute=random_dimension,
+                                     split_value=p, depth=current_depth)
+
+    def update_mass(self, X, is_reference_window):
+        """
+        Records mass data in the tree.
+
+        Parameters
+        ----------
+        X: numpy.ndarray of shape (1, n_features)
+            Instance attributes.
+        is_reference_window: Boolean
+            True if the classifier is in the first reference window, otherwise false
+        Returns
+        -------
+
+        """
+        return self._update_mass(X, self.root, is_reference_window)
+
+    def _update_mass(self, X, node, is_reference_window):
+        """
+        Private function where actual mass data is recorded.
+
+        Parameters
+        ----------
+        X: numpy.ndarray of shape (1, n_features)
+            Instance attributes.
+        node: HalfSpaceTreeNode
+
+        is_reference_window: Boolean
+            True if the classifier is in the first reference window, otherwise false
+
+        """
+        if node is None:
+            return
+
+        if is_reference_window:
+            node.r = node.r + 1
+        else:
+            node.l = node.l + 1
+
+        if node.depth < self.max_depth:
+            if X[node.split_attribute] > node.split_value:
+                self._update_mass(X, node.right, is_reference_window)
+            else:
+                self._update_mass(X, node.left, is_reference_window)
+        else:
+            return
+
+    def update_model(self, node):
+        """
+        Updates model to adapt to new data.
+
+        Parameters
+        ----------
+        node: HalfSpaceTreeNode
+
+        """
+        if node is not None:
+            if node.r is not 0 and node.l is not 0:
```

Comment:
Why use this condition instead of the one in MOA?
```python
if is_reference_window:
```

Reply:
I am citing directly from the paper:

> Update model : 
> 1. Node.r ← Node.l for every node with non-zero mass r or l
> 2. Reset Node.l ← 0 for every node with non-zero mass l

Implementation from MOA:

```
		this.r = this.l;
		this.l = 0;
		if(internalNode)
		{
			left.updateModel();
			right.updateModel();
		}
```

Looking to the algorithm provided, I don't know why we should not update leaves also.

Reply:
I see, thanks for the reference. Two comments:

1. From the paper, should the condition be:
```python
if node.r is not 0 or node.l is not 0:
```
2. It seems to me that both implementations are updating all nodes (including leaf nodes). In MOA, the condition `if(internalNode)` is used to stop the update, whereas your implementation has the condition `if node is not None` at the begining.

Reply:
1. You are completely right, it should be an `or`. 

1. Yes I think I was considering another term when I was writing the comment, both implementations are updating leaves. I think I was wanted to be cautious and used `if node is not None:` instead of `if is_reference_window:`. Since we removed `if node is not None:` in the other parts of the code and I can do the same here and replace it with `if is_reference_window:` ?

Reply:
~I vote to remove it. As far as I can tell there is no reason to expect such a condition.~
Looking back into this, we need the condition to stop the recursive update (in MOA is done via the internalNode check)